pos_blstm/blstm.py

```diff
@@ -0,0 +1,206 @@
+from __future__ import print_function
+from builtins import range
+from os.path import dirname, join
+import autograd.numpy as np
+import autograd.numpy.random as npr
+from autograd import grad
+from autograd.scipy.misc import logsumexp
+import sys
+from optimizers import adam
+from rnn import string_to_one_hot, one_hot_to_string,\
+                build_dataset, sigmoid, concat_and_multiply
+import pdb
+import cPickle as pickle
+
+def init_lstm_params(input_size, state_size, output_size,
+                     param_scale=0.01, rs=npr.RandomState(0)):
+    # state_size is length of h vector, input_size is length of x vector
+    # output_size is length of y vector?
+    def rp(*shape): # returns a np array sampled from normal distribution of specified shape
+        return rs.randn(*shape) * param_scale
+
+    if load:
+        return dict(np.load('tmp/trained_weights.npz'))
+    return {'capitals':     rp(3, input_size),
+            'init cells1':   rp(1, state_size),
+            'init hiddens1': rp(1, state_size),
+            'change1':       rp(input_size + state_size + 1, state_size),
+            'forget1':       rp(input_size + state_size + 1, state_size),
+            'ingate1':       rp(input_size + state_size + 1, state_size),
+            'outgate1':      rp(input_size + state_size + 1, state_size),
+            'init cells2':   rp(1, state_size),
+            'init hiddens2': rp(1, state_size),
+            'change2':       rp(input_size + state_size + 1, state_size),
+            'forget2':       rp(input_size + state_size + 1, state_size),
+            'ingate2':       rp(input_size + state_size + 1, state_size),
+            'outgate2':      rp(input_size + state_size + 1, state_size),
+            'predict':      rp(state_size*2 + 1, output_size)}
+
+def lstm_predict(params, inputs, capitals):
+    def update_forward_lstm(_input, hiddens, cells):
+        change  = np.tanh(concat_and_multiply(params['change1'], _input, hiddens))
+        forget  = sigmoid(concat_and_multiply(params['forget1'], _input, hiddens))
+        ingate  = sigmoid(concat_and_multiply(params['ingate1'], _input, hiddens))
+        outgate = sigmoid(concat_and_multiply(params['outgate1'], _input, hiddens))
+        cells   = cells * forget + ingate * change
+        hiddens = outgate * np.tanh(cells)
+        return hiddens, cells
+
+    def update_backward_lstm(_input, hiddens, cells):
+        change  = np.tanh(concat_and_multiply(params['change2'], _input, hiddens))
+        forget  = sigmoid(concat_and_multiply(params['forget2'], _input, hiddens))
+        ingate  = sigmoid(concat_and_multiply(params['ingate2'], _input, hiddens))
+        outgate = sigmoid(concat_and_multiply(params['outgate2'], _input, hiddens))
+        cells   = cells * forget + ingate * change
+        hiddens = outgate * np.tanh(cells)
+        return hiddens, cells
+
+    def hiddens_to_output_probs(hiddens):
+        output = concat_and_multiply(params['predict'], hiddens)
```

Comment:
I'm not 100% sure about this but I think you need to feed both forward and backward to the output layer, if I only used to forward pass to classify then wouldn't it not consider any information about the reverse ordering?